internal/provider/resource_model.go

```diff
@@ -7,326 +7,522 @@ import (
 
-func resourceModelRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+	r.trace(fmt.Sprintf("model resource created: %q", modelName))
 
-	var diags diag.Diagnostics
+	// Write the state plan into the Response.State
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+}
 
-	uuid := d.Id()
-	response, err := client.Models.ReadModel(uuid)
-	if err != nil {
-		return handleModelNotFoundError(err, d)
+func (r *modelResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	// Prevent panic if the provider has not been configured.
+	if r.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "model", "read")
+		return
 	}
 
-	cloudList := []map[string]interface{}{{
-		"name":   strings.TrimPrefix(response.ModelInfo.CloudTag, juju.PrefixCloud),
-		"region": response.ModelInfo.CloudRegion},
-	}
+	var state modelResourceModel
 
-	if err := d.Set("name", response.ModelInfo.Name); err != nil {
-		return diag.FromErr(err)
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	if err := d.Set("cloud", cloudList); err != nil {
-		return diag.FromErr(err)
+	uuid := state.ID.ValueString()
+	response, err := r.client.Models.ReadModel(uuid)
+	if err != nil {
+		resp.Diagnostics.Append(handleModelNotFoundError(ctx, err, &resp.State)...)
+		return
 	}
-	if err := d.Set("constraints", response.ModelConstraints.String()); err != nil {
-		return diag.FromErr(err)
+	r.trace(fmt.Sprintf("found model: %v", uuid))
+
+	// Acquire cloud, credential, and config
+	cloudList := []nestedCloud{{
+		Name:   types.StringValue(strings.TrimPrefix(response.ModelInfo.CloudTag, juju.PrefixCloud)),
+		Region: types.StringValue(response.ModelInfo.CloudRegion),
+	}}
+	cloudType := req.State.Schema.GetBlocks()["cloud"].(schema.ListNestedBlock).NestedObject.Type()
+	newStateCloud, errDiag := types.ListValueFrom(ctx, cloudType, cloudList)
+	resp.Diagnostics.Append(errDiag...)
+	if resp.Diagnostics.HasError() {
+		return
 	}
+
 	tag, err := names.ParseCloudCredentialTag(response.ModelInfo.CloudCredentialTag)
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to parse cloud credential tag for model, got error: %s", err))
+		return
 	}
 	credential := tag.Name()
-	if err := d.Set("credential", credential); err != nil {
-		return diag.FromErr(err)
-	}
-	if err := d.Set("type", response.ModelInfo.Type); err != nil {
-		return diag.FromErr(err)
-	}
 
 	// Only read model config that is tracked in Terraform
-	config := d.Get("config").(map[string]interface{})
-	for k := range config {
+	var stateConfig map[string]string
+	resp.Diagnostics.Append(state.Config.ElementsAs(ctx, &stateConfig, false)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	for k := range stateConfig {
```

Comment:
As the stateConfig struct is now a `map[string]string{}`, is this conversion loop still necessary?